sw/device/lib/dif/dif_pinmux.c

```diff
@@ -0,0 +1,147 @@
+  for (int i = 0; i < PINMUX_PERIPH_INSEL_MULTIREG_COUNT; ++i) {
+    ptrdiff_t register_offset =
+        PINMUX_PERIPH_INSEL0_REG_OFFSET + (i * PINMUX_REGISTER_WIDTH_IN_BYTES);
+    mmio_region_write32(pinmux->base_addr, register_offset,
+                        PINMUX_INSEL_RESET_DEFAULT);
+  }
+
+  // Reset all of the outsel registers (multireg) to the reset value.
+  for (int i = 0; i < PINMUX_MIO_OUTSEL_MULTIREG_COUNT; ++i) {
+    ptrdiff_t register_offset =
+        PINMUX_MIO_OUTSEL0_REG_OFFSET + (i * PINMUX_REGISTER_WIDTH_IN_BYTES);
+    mmio_region_write32(pinmux->base_addr, register_offset,
+                        PINMUX_OUTSEL_RESET_DEFAULT);
+  }
+}
+
+dif_pinmux_init_error_t dif_pinmux_init(mmio_region_t base_addr,
+                                        dif_pinmux_t *pinmux) {
+  if (pinmux == NULL) {
+    return kDifPinmuxInitErrorInvalidArgument;
+  }
+
+  // Check whether the peripheral can be configured.
+  bool write_enabled = mmio_region_get_bit32(base_addr, PINMUX_REGEN_REG_OFFSET,
+                                             PINMUX_REGEN_WEN);
+  if (!write_enabled) {
+    return kDifPinmuxInitErrorRegistersLocked;
+  }
+
+  pinmux->base_addr = base_addr;
+  pinmux->write_enabled = write_enabled;
+
+  // Set the peripheral to the reset defaults.
+  dif_pinmux_reset(pinmux);
+
+  return kDifPinmuxInitErrorNone;
+}
+
+dif_pinmux_registers_lock_error_t dif_pinmux_registers_lock(
+    dif_pinmux_t *pinmux) {
+  if (pinmux == NULL) {
+    return kDifPinmuxRegistersLockErrorInvalidArgument;
+  }
+
+  mmio_region_write32(pinmux->base_addr, PINMUX_REGEN_REG_OFFSET, 0);
+  pinmux->write_enabled = false;
+
+  return kDifPinmuxRegistersLockErrorNone;
+}
+
+dif_pinmux_insel_set_error_t dif_pinmux_insel_set(
+    const dif_pinmux_t *pinmux, dif_pinmux_peripheral_in_t peripheral_input,
+    dif_pinmux_insel_t mio_select) {
+  if (pinmux == NULL) {
+    return kDifPinmuxInselSetErrorInvalidArgument;
+  }
+
+  if (!pinmux->write_enabled) {
+    return kDifPinmuxInselSetErrorRegistersLocked;
+  }
+
+  if (mio_select < kDifPinmuxInselTieToZero ||
+      mio_select > kDifPinmuxInselPadLast) {
+    return kDifPinmuxInselSetErrorBadMioSelect;
+  }
+
+  // Register offset based on a given peripheral input and MIO select.
+  uint32_t register_index =
+      peripheral_input / PINMUX_PERIPH_INSEL_IN_FIELDS_PER_REG;
+  ptrdiff_t register_offset = PINMUX_PERIPH_INSEL0_REG_OFFSET +
+                              (register_index * PINMUX_REGISTER_WIDTH_IN_BYTES);
+
+  uint32_t field_index =
+      peripheral_input % PINMUX_PERIPH_INSEL_IN_FIELDS_PER_REG;
+  uint32_t field_offset = field_index * PINMUX_PERIPH_INSEL_IN_FIELD_WIDTH;
+
+  // INSEL field masks are the same for all of the fields.
+  bitfield_field32_t field = {
+      .mask = PINMUX_PERIPH_INSEL0_IN0_MASK,
+      .index = field_offset,
+      .value = mio_select,
+  };
+  mmio_region_nonatomic_set_field32(pinmux->base_addr, register_offset, field);
```

Comment:
This is really ergonomic and a pleasure to read! Thank you!

Reply:
Thanks :)